### PR TITLE
Add `ScaleInvariantError` Loss

### PR DESCRIPTION
**Pull request by awsaf49**
This PR will add `ScaleInvariantError (SIE)`  loss as mentioned in this issue, #17719

### CODE REVIEW

**Comment by awsaf49** [comment]
@qlzh727 Hi, 
While writing test cases for loss, I observed an issue with the mask function and its impact on `ReductionV2`. Specifically, I noticed that the use of mask (to avoid invalid depth) caused a loss of shape in y_true and y_pred, ultimately converting them into `1D` vectors. Additionally, since the loss function contains `log`, multiplying the mask to zero values is not a straightforward solution.

In the context of Depth Estimation, it is essential to consider the occurrence of invalid depths (i.e., depth = 0) due to sensor limitations. In fact, two of the most popular datasets used for Depth Estimation, KITTI and NYU-Depth-V2, contain instances of invalid depth.

Also not sure if we can avoid mask as in Depth Estimation, `invalid depths ( depth = 0)` are very likely to happen due to sensor. Two most popular dataset **KITTI** and **NYU-Depth-V2** contains invalid depth.

Any thoughts on what should we do?

**Review by rchao — CHANGES_REQUESTED** [review]
Thanks for the PR! Can you also add unit test to cover the behavior of the loss class/function?

File: keras/losses.py [hunk 1]
```diff
@@ -419,6 +419,61 @@ def __init__(
         super().__init__(mean_absolute_error, name=name, reduction=reduction)
 
 
+@keras_export("keras.losses.ScaleInvariantError")
+class ScaleInvariantError(LossFunctionWrapper):
+    """Computes the Scale Invariant Error between labels and predictions.
+
+    Formulas:
+
+    ```
```
Comment (fchollet) [inline keras/losses.py:428]: Add: `python`

File: keras/losses.py [hunk 2]
```diff
@@ -419,6 +419,61 @@ def __init__(
...
+    g = log(y_pred) - log(y_true)
+    loss = alpha * sqrt(mean(g ** 2) - lamda * mean(g) ** 2)
+    ```
+
+    Standalone usage:
```
Comment (fchollet) [inline keras/losses.py:433]: `Example:`

File: keras/losses.py [hunk 3]
```diff
@@ -419,6 +419,61 @@ def __init__(
...
+    def __init__(
+        self,
+        reduction=losses_utils.ReductionV2.AUTO,
+        alpha=10.0,
+        lamda=0.85,
+        name="scale_invariant_error",
+    ):
+        """Initializes `ScaleInvariantError` instance.
+
+        Args:
```
Comment (fchollet) [inline keras/losses.py:452]: Please fix indent (4 spaces)

Also: please move the Args section to the class docstring and remove this docstring.

File: keras/losses.py [hunk 4]
```diff
@@ -1755,6 +1810,61 @@ def _ragged_tensor_mae(y_true, y_pred):
     return _ragged_tensor_apply_loss(mean_absolute_error, y_true, y_pred)
 
 
+@keras_export(
+    "keras.metrics.scale_invariant_error",
```
Comment (fchollet) [inline keras/losses.py:1814]: Please do not export it as a metric for the time being.

File: keras/losses.py [hunk 5]
```diff
@@ -1755,6 +1810,61 @@ def _ragged_tensor_mae(y_true, y_pred):
...
+    Standalone usage:
+
+    >>> y_true = [[1, 2], [3, 4]]
+    >>> y_pred = [[2, 4], [6, 8]]
+    >>> loss = tf.keras.losses.scale_invariant_error(y_true, y_pred)
+    >>> assert loss.shape == (2,)
+    >>> loss.numpy()
+    array([1.6659865, 1.6659865], dtype=float32)
+
+    Args:
```
Comment (fchollet) [inline keras/losses.py:1838]: Please fix indent (4 spaces)

File: keras/losses.py [hunk 6]
```diff
@@ -1755,6 +1810,61 @@ def _ragged_tensor_mae(y_true, y_pred):
...
+)
+@tf.__internal__.dispatch.add_dispatch_support
+def scale_invariant_error(y_true, y_pred, alpha=10.0, lamda=0.85):
+    """Computes the Scale Invariant Error between labels and predictions.
+
+    According to [Eigen et al., 2014](https://arxiv.org/pdf/1406.2283), it
+    measures the error by comparing relationships between pairs of pixels
+    between points in the scene, irrespective of the absolute global scale.
+
+    Standalone usage:
```
Comment (fchollet) [inline keras/losses.py:1829]: `Example:`

**Review by fchollet — COMMENTED** [review]
Thank you for the PR! Looks good!

**Comment by awsaf49** [comment]
@fchollet @rchao could you please check?

**Comment by awsaf49** [comment]
@divyashreepathihalli any update?

**Comment by sachinprasadhs** [comment]
Hello, Thank you for submitting a pull request.

We're currently in the process of migrating the new `Keras 3 code base` from `keras-team/keras-core` to `keras-team/keras`. 
Consequently, merging this PR is not possible at the moment. After the migration is successfully completed, feel free to reopen this PR at `keras-team/keras` if you believe it remains relevant to the `Keras 3 code base`. If instead this PR fixes a bug or security issue in legacy `tf.keras`, you can instead reopen the PR at `keras-team/tf-keras`, which hosts the `TensorFlow-only`, legacy version of `Keras`.
